x-pack/plugins/maps/public/store/map.js

```diff
@@ -359,3 +363,57 @@ function getValidDataRequest(state, action, checkRequestToken = true) {
 function findLayerById(state, id) {
   return state.layerList.find(layer => layer.id === id);
 }
+
+function rollbackAnyTrailingTrackedLayerState(state) {
+  state.layerList.forEach(layer => {
+    if (layer[TRACKED_LAYER_DESCRIPTOR]) {
+      rollbackTrackedLayerState(state, layer.id);
+    }
+  });
+}
+
+function trackCurrentLayerState(state, layerId) {
+  const layer = findLayerById(state, layerId);
+  const layerCopy = copyPersistentState(layer);
+  return updateLayerInList(state, layerId, TRACKED_LAYER_DESCRIPTOR, layerCopy);
+}
+
+function removeTrackedLayerState(state, layerId) {
+  const layer = findLayerById(state,  layerId);
+  if (!layer) {
+    return state;
+  }
+
+  const layerIndex = getLayerIndex(state.layerList, layerId);
+  const copyLayer = { ...layer };
+  delete copyLayer[TRACKED_LAYER_DESCRIPTOR];//just update in place
+  const newLayerList = [...state.layerList];
+  newLayerList[layerIndex] = copyLayer;
+  return {
+    ...state,
+    layerList: newLayerList
+  };
+}
+
+function rollbackTrackedLayerState(state, layerId) {
+  const layer = findLayerById(state, layerId);
+  if (!layer) {
+    return state;
+  }
+  const trackedLayerDescriptor = layer[TRACKED_LAYER_DESCRIPTOR];
+
+  //this assumes that any nested temp-state in the layer-descriptor (e.g. of styles), is not relevant and can be recovered easily (e.g. this is not the case for __dataRequests)
+  //That assumption is true in the context of this app, but not generalizable.
+  //consider rewriting copyPersistentState to only strip the first level of temp state.
+  const rolledbackLayer = { ...layer, ...trackedLayerDescriptor };
+  delete rolledbackLayer[TRACKED_LAYER_DESCRIPTOR];
+
+  const layerIndex = getLayerIndex(state.layerList, layerId);
+  state.layerList[layerIndex] = rolledbackLayer;
```

Comment:
Remove the line `state.layerList[layerIndex] = rolledbackLayer`. It is modifying original state